src/python/pants/engine/target.py

```diff
@@ -2675,7 +2679,40 @@ class DescriptionField(StringField):
     )
 
 
-COMMON_TARGET_FIELDS = (Tags, DescriptionField)
+class VisibilityField(AsyncFieldMixin, StringSequenceField):
+    PUBLIC_VISIBILITY = "<public>"
+    PRIVATE_VISIBILITY = "<private>"
+
+    alias = "visibility"
+    default = (PUBLIC_VISIBILITY,)
+    help = softwrap(
+        """
+        The `visibility` value specifies a set of project paths, and only targets from within one of
+        these paths are allowed to depend on this target.
+
+        There are two special values, `"<public>"` and `"<private>"`.
+
+        A `"<public>"` visibility will allow any target to depend on this target, while
+        `"<private>"` visibility only allows targets in the same subtree as this target.
+
+        """
+    )
+
+    def visible(self, to_address: Address) -> bool:
+        # Add trailing slash to cut off paths with common prefixes but different suffixes.
+        target_path = self.address.spec_path + "/"
+        to_path = to_address.spec_path + "/"
+        for rule in self.value or ():
+            if rule == self.PUBLIC_VISIBILITY:
+                return True
+            if rule == self.PRIVATE_VISIBILITY:
+                return to_path.startswith(target_path)
+            if to_path.startswith(rule):
```

Comment:
I think that this would need to validate that `rule` ends in a slash as well?

I'd suggest switching both of these to https://github.com/pantsbuild/pants/blob/b98649e8f64c99f85d5fd5017f6f117609d5ac01/src/python/pants/util/dirutil.py#L44-L48, which avoids the need for the trailing slash.

Reply:
Switching the entire logic to instead use `Specs` for determining valid targets.